File: Practice/atcoder/ABC/054/src/c.rb
# dfs使えばいいんじゃないかな
n, m = gets.chomp.split.map(&:to_i)
graph = Array.new(n).map{[]}
m.times do
  a, b = gets.chomp.split.map(&:to_i)
  graph[a-1] << b-1
  graph[b-1] << a-1
end
# p graph
Track = Struct.new(:nowpath, :visited)

paths = [Track.new(graph[0], [0])]
ans = 0
while paths.size > 0 do
  cur = paths.pop
  ans += cur.visited.size == n ? 1 : 0
  # puts "next available: #{cur.nowpath}"
  cur.nowpath.each do |next_id|
    unless cur.visited.include?(next_id)
      # p cur.visited
      next_track = Track.new(graph[next_id], cur.visited.clone << next_id)
      paths << next_track
    end
  end
end

puts ans

File: Practice/atcoder/ABC/054/src/c.py
# dfsでみる どうせ小さいので都度配列を渡していい
# 隣接リストでもつことにする
# クラスで管理するのすらめんどくさいのでlistオブジェクトでいいや

n, m = map(int, input().split())
graph = [[] for _ in range(n)]
for _ in range(m):
    s, t = map(int, input().split())
    graph[s - 1].append(t - 1)
    graph[t - 1].append(s - 1)

ans = 0

path = [0, [0]]
stack = [path]

while len(stack) > 0:
    cur = stack.pop()
    if len(cur[1]) == n:
        ans += 1
    else:
        for g in graph[cur[0]]:
            if g not in cur[1]:
                nex = cur[1][:]
                nex.append(g)
                stack.append([g, nex])

print(ans)
